be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala

```diff
@@ -0,0 +1,102 @@
+      Monitor.props(testProbe.ref, heartbeatRate = fastRate, messageDelay = fastRate)
+    )
+
+    testProbe.send(monitorRef, Monitor.AtLeastOneServerConnected)
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+  }
+
+  test("monitor should schedule single heartbeat when receiving a Right GraphMessage") {
+
+    val testProbe = TestProbe()
+    val monitorRef = system.actorOf(
+      Monitor.props(testProbe.ref, heartbeatRate = slowRate, messageDelay = fastRate)
+    )
+
+    // Needed to tell monitor ConnectionMediatorRef
+    val sink = Monitor.sink(monitorRef)
+    testProbe.send(monitorRef, Monitor.AtLeastOneServerConnected)
+
+    Source.single(Right(JsonRpcRequestExample.subscribeRpcRequest)).to(sink).run()
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+  }
+
+  test("monitor should not schedule single hearbeats when receiving a heartbeat or get_messages_by_id") {
+
+    val heartbeat = JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.HEARTBEAT, new ParamsWithMap(Map.empty), None)
+    val getMessagesById = JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.GET_MESSAGES_BY_ID, new ParamsWithMap(Map.empty), None)
+
+    val testProbe = TestProbe()
+    val monitorRef = system.actorOf(
+      Monitor.props(testProbe.ref, heartbeatRate = slowRate, messageDelay = fastRate)
+    )
+
+    // Needed to tell monitor ConnectionMediatorRef
+    val sink = Monitor.sink(monitorRef)
+    testProbe.send(monitorRef, Monitor.AtLeastOneServerConnected)
+    Source.single(Right(heartbeat)).to(sink).run()
+
+    testProbe.expectNoMessage(timeout)
+
+    // Check that monitor is still doing fine when receiving other Right(...)
+    Source.single(Right(JsonRpcRequestExample.subscribeRpcRequest)).to(sink).run()
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+
+    Source.single(Right(getMessagesById)).to(sink).run()
+    testProbe.expectNoMessage(timeout)
+  }
+
+  test("monitor should send heartbeats only when servers are connected") {
+
+    val testProbe = TestProbe()
+    val monitorRef = system.actorOf(
+      Monitor.props(testProbe.ref, heartbeatRate = fastRate, messageDelay = fastRate)
+    )
+
+    // Needed to tell monitor ConnectionMediatorRef
+    testProbe.send(monitorRef, Monitor.AtLeastOneServerConnected)
+
+    // Wait for the first heartbeat then "disconnect" servers
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+    testProbe.send(monitorRef, Monitor.NoServerConnected)
+    testProbe.expectNoMessage(timeout)
+
+    // No single heartbeat should be scheduled either
+    Source.single(Right(JsonRpcRequestExample.subscribeRpcRequest)).to(Monitor.sink(monitorRef)).run()
+    testProbe.expectNoMessage(timeout)
+
+    // Connect a server and check for heartbeats again
```

Comment:
I do not think that those last lines do really belong to the corresponding test. 
You already tested it in previous tests, and it makes the test a bit confusing. I understand that the purpose was to test corner cases, but maybe writing another test by specifying in the description that you want to test if by connecting and disconnecting the servers, the Monitor still behaves correctly.

Reply:
The test goes as follows: 

`monitor should send heartbeats only when servers are connected`

1. Initiate monitor by telling it about connected servers
2. expect a heartbeat
3. disconnect the servers
4. expect no heartbeat
5. connect servers again 
6. expect heartbeats

Excluding these last steps 5 and 6 make the test look incomplete to me (what if monitor has a bug that prevent it to start again after some servers came back). I want the test to guarantee that monitor does its job _whenever and only when servers are connected_.  And it looks redundant to have a second test just doing that last bit.

Do you think the name : 
`monitor should send heartbeats if and only if servers are connected` 

Would make it clearer ?

Reply:
Okay I see your point and I agree that the second test would be redundant!
The new name could be a bit clearer in my opinion. :)